### PR TITLE
Feature/frontend rust

### PR DESCRIPTION
**Pull request by jbis9051**

### CODE REVIEW

File: packages/rust/db/migrations/20220528030155_initial.up.sql [hunk 1]
```diff
@@ -40,6 +41,13 @@ CREATE TABLE client (
     created TIMESTAMP DEFAULT CURRENT_TIMESTAMP
 );
 
+ALTER TABLE
+    "user"
+ADD
+    CONSTRAINT fk_user_primary_client_id FOREIGN KEY (primary_client_id) REFERENCES client (id) ON DELETE
+SET
+    NULL;
+
```
Comment (gannonsmith) [inline packages/rust/db/migrations/20220528030155_initial.up.sql:50]: Why is this done instead of initializing the fk the same way every other fk is?

Comment (gannonsmith) [reply packages/rust/db/migrations/20220528030155_initial.up.sql:50]: Do models need to be updated?

File: packages/rust/frontend/migrations/account/20230524013755_initial.up.sql [hunk 2]
```diff
@@ -1,39 +1,56 @@
...
     identity TEXT NOT NULL,
-    updated_date INTEGER NOT NULL
+    updated_date DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP
 );
 
 CREATE TABLE client (
     id INTEGER PRIMARY KEY AUTOINCREMENT,
     uuid TEXT NOT NULL UNIQUE,
-    user_id INTEGER NOT NULL REFERENCES user (id) ON DELETE CASCADE,
+    user_uuid TEXT NOT NULL,
```
Comment (gannonsmith) [inline packages/rust/frontend/migrations/account/20230524013755_initial.up.sql:13]: Foreign key?

Comment (gannonsmith) [reply packages/rust/frontend/migrations/account/20230524013755_initial.up.sql:13]: On Delete Cascade?

File: packages/rust/frontend/migrations/account/20230524013755_initial.up.sql [hunk 3]
```diff
@@ -1,39 +1,56 @@
...
-    validated_date INTEGER NULL
+    validated_date INTEGER NULL,
+    created_date DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP
 );
 
 CREATE TABLE location (
     id INTEGER PRIMARY KEY AUTOINCREMENT,
-    client_id INTEGER NOT NULL REFERENCES client (id) ON DELETE CASCADE,
+    client_uuid TEXT NOT NULL,
+    group_uuid TEXT NOT NULL,
```
Comment (gannonsmith) [inline packages/rust/frontend/migrations/account/20230524013755_initial.up.sql:22]: Foreign Keys?

File: packages/rust/frontend/migrations/account/20230524013755_initial.up.sql [hunk 4]
```diff
@@ -1,39 +1,56 @@
...
     longitude REAL NOT NULL,
     latitude REAL NOT NULL,
-    raw TEXT NOT NULL,
-    created_date INTEGER NOT NULL,
-    group_uuid TEXT NOT NULL
+    location_date INTEGER NOT NULL,
```
Comment (gannonsmith) [inline packages/rust/frontend/migrations/account/20230524013755_initial.up.sql:25]:
```suggestion
    location_date DATETIME NOT NULL,
```

File: packages/rust/frontend/migrations/account/20230524013755_initial.up.sql [hunk 5]
```diff
@@ -1,39 +1,56 @@
...
+    "key" BLOB NOT NULL UNIQUE,
+    value BLOB NOT NULL UNIQUE,
+    type_name  TEXT NOT NULL,
+    created_date DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP
+);
+
+CREATE TABLE inbox (
+    id INTEGER PRIMARY KEY AUTOINCREMENT,
+    message BLOB NOT NULL,
+    received_date DATETIME NOT NULL
```
Comment (gannonsmith) [inline packages/rust/frontend/migrations/account/20230524013755_initial.up.sql]:
```suggestion
    received_date DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP
```

File: packages/rust/frontend/src/api/client.rs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                self.domain, client_uuid
+            ))
+            .send()
+            .await?
+            .error_for_status()?
+            .json()
+            .await?;
+        let key_package = key_package.key_package;
+        let key_package_in = KeyPackageIn::tls_deserialize(&mut key_package.as_slice()).unwrap();
+        Ok(key_package_in)
```
Comment (gannonsmith) [inline packages/rust/frontend/src/api/client.rs:29]: condense

File: packages/rust/frontend/src/api/client.rs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                signature: Base64(signature),
+            })
+            .send()
+            .await?
+            .error_for_status()?
+            .json()
+            .await?;
+        Ok(res.client_uuid)
+    }
+}
```
Comment (gannonsmith) [inline packages/rust/frontend/src/api/client.rs:85]: Missing PATCH and DELETE for client

File: docs/API_ROUTES.md
Comment (gannonsmith) [inline docs/API_ROUTES.md:1]: Consider adding "/v1/" to api routes

File: packages/rust/frontend/src/api/user.rs [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+            .error_for_status()?
+            .json()
+            .await?;
+        Ok(())
+    }
+
+    pub async fn forgot_check(&self, token: Uuid) -> Result<bool, reqwest::Error> {
+        let res = self
+            .client
+            .get(&format!("{}/v1/user/reset/{}", self.domain, token))
```
Comment (gannonsmith) [inline packages/rust/frontend/src/api/user.rs:131]:
```suggestion
            .get(&format!("{}/v1/user/reset?", self.domain))
```

Comment (gannonsmith) [reply packages/rust/frontend/src/api/user.rs:131]: Not sure exactly how .query() works with an object

File: packages/rust/frontend/src/api/user.rs [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+            .query(&PasswordResetCheck { token })
+            .send()
+            .await?;
+        Ok(res.status() != StatusCode::NOT_FOUND)
+    }
+
+    pub async fn search(&self, query: String) -> Result<Vec<PublicUser>, reqwest::Error> {
```
Comment (gannonsmith) [inline packages/rust/frontend/src/api/user.rs:138]: Missing documentation for /v1/user/search in API docs

File: packages/rust/frontend/src/application_message.rs [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+pub enum Message {
+    Location(Location),
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+#[bridge]
+pub struct Location {
+    pub latitude: f64,
+    pub longitude: f64,
+    pub timestamp: i64,
```
Comment (gannonsmith) [inline packages/rust/frontend/src/application_message.rs:16]: surely there is a better type for a timestamp?

File: packages/rust/frontend/src/helper/bubble_group.rs [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    pub fn new_from_uuid(
+        group_uuid: &Uuid,
+        mls_provider: &impl OpenMlsCryptoProvider,
+    ) -> Option<Self> {
+        MlsGroup::load(&GroupId::from_slice(group_uuid.as_ref()), mls_provider).map(Self::new)
+    }
+
+    pub fn group_uuid(&self) -> Uuid {
+        self.group_uuid
+    }
```
Comment (gannonsmith) [inline packages/rust/frontend/src/helper/bubble_group.rs:35]: Why does this exist?

File: packages/rust/frontend/src/helper/bubble_group.rs [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+                index: member.index,
+                user_uuid,
+                client_uuid,
+            });
+        }
+        Ok(client_uuids)
+    }
+
+    pub fn save_if_needed(&mut self, mls_provider: &MlsProvider) -> Result<(), Error> {
+        if matches!(self.group.state_changed(), InnerState::Changed) {
```
Comment (gannonsmith) [inline packages/rust/frontend/src/helper/bubble_group.rs:77]: Is matches! macro needed?

File: packages/rust/frontend/src/helper/helper.rs [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+use crate::Error;
+
+pub async fn get_this_client_mls_resources(
+    user_uuid: &Uuid,
+    client_uuid: &Uuid,
+    account_db: &DbPool,
+    mls_provider: &MlsProvider,
+) -> Result<(SignatureKeyPair, CredentialWithKey), Error> {
+    let client_public = base64::deserialize(
+        // TODO: consider storing this directly in GlobalAccountData
```
Comment (gannonsmith) [inline packages/rust/frontend/src/helper/helper.rs:20]: Consider

File: packages/rust/frontend/src/helper/helper.rs [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        credential,
+        signature_key: SignaturePublicKey::from(signature.public()),
+    };
+    Ok((signature, credential_with_key))
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ParseIdentityError {}
+
+pub fn parse_identity(identity: &[u8]) -> Result<(Uuid, Uuid), ParseIdentityError> {
```
Comment (gannonsmith) [inline packages/rust/frontend/src/helper/helper.rs:40]: Move ParseIdentityError to Error enum in frontend/src/lib.rs

File: packages/rust/frontend/src/helper/resource_fetcher.rs [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+        user_uuid: &Uuid,
+    ) -> Result<Vec<PublicClient>, ResourceError> {
+        let user = self.get_user_full_authentication(user_uuid).await?;
+        let clients = self.api.get_user_clients(user_uuid).await?;
+        self.authenticate_clients_against_user_identity(&user.identity, &clients)?;
+
+        for client in clients.clone() {
+            let mut client: Client = client.into();
+            Client::delete_by_uuid(&self.account_db, &client.uuid).await?;
+            client.create(&self.account_db).await?;
```
Comment (gannonsmith) [inline packages/rust/frontend/src/helper/resource_fetcher.rs:58]: Race Condition? Maybe we should add model for "try_create"?

File: packages/rust/frontend/src/helper/resource_fetcher.rs [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+    pub async fn get_user_partial_authentication(
+        &self,
+        user_uuid: &Uuid,
+    ) -> Result<User, ResourceError> {
+        let local_user = User::try_from_uuid(&self.account_db, user_uuid).await?;
+        if let Some(cache_user) = local_user {
+            return Ok(cache_user);
+        }
+        let api_user = self.api.get_user(user_uuid).await?;
+        let mut user: User = api_user.into();
```
Comment (gannonsmith) [inline packages/rust/frontend/src/helper/resource_fetcher.rs:112]:
```suggestion
        let mut user: User = self.api.get_user(user_uuid).await?.into();
```

File: packages/rust/frontend/src/js_interface/client.rs [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+        let mls_provider = MlsProvider::new(account_db.clone());
+        let (signature, _) =
+            get_this_client_mls_resources(user_uuid, &client_uuid, account_db, &mls_provider)
+                .await?;
+
+        let identity = format!("client_{}_{}", user_uuid, client_uuid);
+        let credential = Credential::new(identity.into_bytes(), CredentialType::Basic)?;
+        let public = SignaturePublicKey::from(signature.public());
+
+        let num_key_packages = 100;
```
Comment (gannonsmith) [inline packages/rust/frontend/src/js_interface/client.rs:32]: why 100?

File: packages/rust/frontend/src/js_interface/group.rs [hunk 18]
```diff
@@ -0,0 +1,292 @@
...
+        let old_members = group
+            .get_group_members()?
+            .into_iter()
+            .filter(|m| m.client_uuid != client_uuid) // we don't want to send the message to ourselves
+            .map(|m| m.client_uuid)
+            .collect::<Vec<_>>();
+
+        let (mls_message_out, welcome_out, _group_info) =
+            group.add_members(&mls_provider, &signature, &key_packages)?;
+        // TODO what happens if we add a member that is already in the group?
```
Comment (gannonsmith) [inline packages/rust/frontend/src/js_interface/group.rs:152]: Can we just not permit this in frontend?

File: packages/rust/frontend/src/js_interface/group.rs [hunk 19]
```diff
@@ -0,0 +1,292 @@
...
+        api.send_message(old_members, mls_message_out, group_uuid)
+            .await?;
+
+        group.save_if_needed(&mls_provider)?;
+
+        Ok(())
+    }
+
+    //#[bridge]
+    pub async fn remove_member(&self, group_uuid: Uuid, user_uuid: Uuid) -> Result<(), Error> {
```
Comment (gannonsmith) [inline packages/rust/frontend/src/js_interface/group.rs:172]: should we make this remove_members() and pass a Vec of user_uuids instead?

File: packages/rust/frontend/src/js_interface/location.rs [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+            .map(|location| Location {
+                longitude: location.longitude,
+                latitude: location.latitude,
+                timestamp: location.location_date.timestamp_millis(),
+            })
+            .collect())
+    }
+
+    #[bridge]
+    pub async fn get_num_location(
```
Comment (gannonsmith) [inline packages/rust/frontend/src/js_interface/location.rs:41]:
```suggestion
    pub async fn get_num_locations(
```

File: packages/rust/frontend/src/js_interface/message.rs [hunk 21]
```diff
@@ -0,0 +1,288 @@
...
+                        .unwrap();
+                    }
+
+                    group.save_if_needed(&mls_provider).unwrap();
+                }
+                MlsMessageInBody::GroupInfo(_) => continue,
+                MlsMessageInBody::KeyPackage(_) => continue,
+            }
+
+            // TODO: process message
```
Comment (gannonsmith) [inline packages/rust/frontend/src/js_interface/message.rs:134]: ?

File: packages/rust/frontend/src/mls_provider/keystore.rs [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+
+impl From<sqlx::Error> for SqlxError {
+    fn from(e: sqlx::Error) -> Self {
+        Self(e)
+    }
+}
+
+impl PartialEq for SqlxError {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.to_string() == other.0.to_string() // TODO: This is not a good way to compare errors
```
Comment (gannonsmith) [inline packages/rust/frontend/src/mls_provider/keystore.rs:31]: true

File: packages/rust/frontend/src/mls_provider/mod.rs [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+        &self.crypto
+    }
+
+    fn rand(&self) -> &Self::RandProvider {
+        &self.crypto
+    }
+
+    fn key_store(&self) -> &Self::KeyStoreProvider {
+        &self.key_store
+    }
```
Comment (gannonsmith) [inline packages/rust/frontend/src/mls_provider/mod.rs:37]: why not just make these attributes public?

File: packages/rust/frontend/src/models/account/group.rs [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+        .bind(&self.image)
+        .fetch_one(db)
+        .await?)
+            .into();
+        Ok(())
+    }
+
+    pub async fn all(db: &DbPool) -> Result<Vec<Group>, sqlx::Error> {
+        let locations = sqlx::query("SELECT * FROM \"group\"").fetch_all(db).await?;
+        Ok(locations.iter().map(Group::from).collect())
```
Comment (gannonsmith) [inline packages/rust/frontend/src/models/account/group.rs:40]:
```suggestion
        Ok(locations.iter().map(|g| g.into()).collect())
```

Comment (gannonsmith) [reply packages/rust/frontend/src/models/account/group.rs:40]: idk if this is better or even works

File: packages/rust/frontend/src/models/account/inbox.rs [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+            message: row.get("message"),
+            received_date: row.get("received_date"),
+        }
+    }
+}
+
+impl Inbox {
+    pub async fn create(&mut self, db: &DbPool) -> Result<(), sqlx::Error> {
+        *self =
+            (&sqlx::query("INSERT INTO inbox (message, received_date) VALUES (?, ?) RETURNING *")
```
Comment (gannonsmith) [inline packages/rust/frontend/src/models/account/inbox.rs:25]:
```suggestion
            (&sqlx::query("INSERT INTO inbox (message, received_date) VALUES (?, ?) RETURNING *;")
```

File: packages/rust/frontend/src/models/account/inbox.rs [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+                .bind(&self.message)
+                .bind(self.received_date)
+                .fetch_one(db)
+                .await?)
+                .into();
+        Ok(())
+    }
+
+    pub async fn all(db: &DbPool) -> Result<Vec<Inbox>, sqlx::Error> {
+        sqlx::query("SELECT * FROM inbox ORDER BY received_date ASC")
```
Comment (gannonsmith) [inline packages/rust/frontend/src/models/account/inbox.rs:35]:
```suggestion
        sqlx::query("SELECT * FROM inbox ORDER BY received_date ASC;")
```

File: packages/rust/frontend/src/models/account/inbox.rs [hunk 27]
```diff
@@ -0,0 +1,48 @@
...
+            .map(|row: SqliteRow| Inbox::from(&row))
+            .fetch_all(db)
+            .await
+    }
+
+    pub async fn delete_by_id(db: &DbPool, id: i32) -> Result<(), sqlx::Error> {
+        sqlx::query("DELETE FROM inbox WHERE id = ?")
```
Comment (gannonsmith) [inline packages/rust/frontend/src/models/account/inbox.rs:42]:
```suggestion
        sqlx::query("DELETE FROM inbox WHERE id = ?;")
```

File: packages/rust/frontend/src/models/account/keystore.rs [hunk 28]
```diff
@@ -0,0 +1,110 @@
...
+            .map(|row| row.map(|row| row.get("value")))
+    }
+
+    pub async fn set(
+        db: &DbPool,
+        key: &[u8],
+        value: &[u8],
+        type_name: InternalMlsEntityId,
+    ) -> Result<(), sqlx::Error> {
+        sqlx::query("DELETE FROM keystore WHERE key = $1; INSERT INTO keystore (key, value, type_name) VALUES ($1,$2, $3);")
```
Comment (gannonsmith) [inline packages/rust/frontend/src/models/account/keystore.rs:94]: nit: Could separate these into two queries but use one transaction

File: packages/rust/frontend/src/models/account/location.rs [hunk 29]
```diff
@@ -0,0 +1,85 @@
...
+    fn from(row: &SqliteRow) -> Self {
+        let location_date: i64 = row.get("location_date");
+        let location_date = NaiveDateTime::from_timestamp_millis(location_date).unwrap();
+        Self {
+            id: row.get("id"),
+            client_uuid: row.get("client_uuid"),
+            group_uuid: row.get("group_uuid"),
+            longitude: row.get("longitude"),
+            latitude: row.get("latitude"),
+            location_date,
```
Comment (gannonsmith) [inline packages/rust/frontend/src/models/account/location.rs:31]: ?

File: packages/rust/frontend/src/models/account/location.rs [hunk 30]
```diff
@@ -0,0 +1,85 @@
...
+            raw: row.get("raw"),
+            created_date: row.get("created_date"),
+        }
+    }
+}
+
+impl Location {
+    pub async fn create(&mut self, db: &DbPool) -> Result<(), Error> {
+        *self = (&sqlx::query("INSERT INTO location (client_uuid, group_uuid, longitude, latitude, location_date, raw) VALUES ($1, $2, $3, $4, $5, $6) RETURNING *")
```
Comment (gannonsmith) [inline packages/rust/frontend/src/models/account/location.rs:40]:
```suggestion
        *self = (&sqlx::query("INSERT INTO location (client_uuid, group_uuid, longitude, latitude, location_date, raw) VALUES ($1, $2, $3, $4, $5, $6) RETURNING *;")
```

File: packages/rust/frontend/src/models/account/location.rs [hunk 31]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    pub async fn query(
+        db: &DbPool,
+        group_uuid: &Uuid,
+        client_uuid: &Uuid,
+        before: &NaiveDateTime,
+        amount: u32,
+    ) -> Result<Vec<Location>, sqlx::Error> {
+        let locations = sqlx::query("SELECT * FROM location WHERE location_date < $1 AND group_uuid = $2 AND client_uuid = $3 ORDER BY location_date DESC LIMIT $4")
```
Comment (gannonsmith) [inline packages/rust/frontend/src/models/account/location.rs:58]:
```suggestion
        let locations = sqlx::query("SELECT * FROM location WHERE location_date < $1 AND group_uuid = $2 AND client_uuid = $3 ORDER BY location_date DESC LIMIT $4;")
```

File: packages/rust/frontend/src/models/account/location.rs [hunk 32]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    pub async fn count_query(
+        db: &SqlitePool,
+        group_uuid: &Uuid,
+        client_uuid: &Uuid,
+        from: &NaiveDateTime,
+        to: &NaiveDateTime,
+    ) -> Result<i64, sqlx::Error> {
+        let count = sqlx::query("SELECT COUNT(*) as count FROM location WHERE location_date BETWEEN $1 AND $2 AND group_uuid = $3 AND client_uuid = $4")
```
Comment (gannonsmith) [inline packages/rust/frontend/src/models/account/location.rs:75]:
```suggestion
        let count = sqlx::query("SELECT COUNT(*) as count FROM location WHERE location_date BETWEEN $1 AND $2 AND group_uuid = $3 AND client_uuid = $4;")
```

File: packages/rust/server/Cargo.toml [hunk 33]
```diff
@@ -1,30 +1,32 @@
...
-name = "bubble"
+name = "server"
 version = "0.1.0"
 edition = "2021"
 
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-sqlx = { version = "0.6", features = ["runtime-tokio-rustls", "postgres", "migrate", "uuid", "chrono"] }
+sqlx = { version = "0.6", features = ["runtime-tokio-rustls", "sqlite", "postgres", "migrate", "uuid", "chrono"] }
```
Comment (gannonsmith) [inline packages/rust/server/Cargo.toml]: sqlite in backend?

File: packages/rust/server/src/models/user.rs [hunk 34]
```diff
@@ -133,4 +168,12 @@ impl User {
             .await?;
         Ok(())
     }
+
+    pub async fn client(&self, db: &DbPool) -> Result<Option<Client>, sqlx::Error> {
```
Comment (gannonsmith) [inline packages/rust/server/src/models/user.rs]:
```suggestion
    pub async fn primary_client(&self, db: &DbPool) -> Result<Option<Client>, sqlx::Error> {
```

**Review by gannonsmith — CHANGES_REQUESTED** [review]
I ignored error handling in the name of progress. Looking forward to reviewing features with less code in the future.

File: packages/rust/frontend/migrations/account/20230524013755_initial.up.sql (hunk 4 above)
Comment (jbis9051) [reply packages/rust/frontend/migrations/account/20230524013755_initial.up.sql:25]: TODO fix this

File: packages/rust/server/Cargo.toml (hunk 33 above)
Comment (jbis9051) [reply packages/rust/server/Cargo.toml]: @gannonsmith you can do

File: packages/rust/server/src/models/user.rs (hunk 34 above)
Comment (jbis9051) [reply packages/rust/server/src/models/user.rs]: refactor this @gannonsmith

**Pull request merged**